docs_src/design/adr/security/0008-Secret-Creation-and-Distribution.md

```diff
@@ -0,0 +1,332 @@
+# Creation and Distribution of Secrets
+
+## Status
+
+** proposed **
+
+## Context
+
+This ADR seeks to clarify and prioritize the secret handling approach taken by EdgeX.
+
+EdgeX microservices need a number of secrets to be created and distributed
+in order to create a functional, secure system.
+Among these secrets are:
+
+- Privileged administrator passwords (such as a database superuser password)
+- Service account passwords (e.g. non-privileged database accounts)
+- PKI private keys
+
+There is a lack of consistency on how secrets are created and distributed to EdgeX microservices,
+and when developers need to add new components to the system,
+it is unclear on what the preferred approach should be.
+
+This document assumes a threat model wherein the EdgeX services are sandboxed
+(such as in a snap or a container) and the host system is trusted,
+and all services running in a single snap share a trust boundary.
+
+### Terms
+
+The following terms will be helpful for understading the subsequent discussion:
+
+- _SECRETSLOC_ is a protected file system path where transient secrets are stored.
+
+  _SECRETSLOC_ is implementation-dependent. 
+  A desirable feature of _SECRETSLOC_ would be that data written here
+  is kept in RAM and is not persisted to storage media.
+  This property is not achieveable in all circumstances.
+
```

Comment:
As an alternative to a file system and something kept in memory with optional persistence, Redis Labs uses Redis instances for this purpose. The runtime overhead of another Redis instance is a couple of MB and storing a few key/value pairs won't change that. Simple shell scripts can return the values.

Reply:
While the Redis Labs approach is good, many open source projects can only consume data from files, making it the lowest-common denominator.

Reply:
@andresrinivasan The way I have explained this is that _SECRETSLOC_ is the equivalent of `169.254.169.254` in the public cloud -- it delivers bootstrapping secrets.